Allow using native c10d_functional via _functional_collectives

yifuwang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #118498
* #118437
* #118423
* #118426
* __->__ #113057

This diff introduces an env var `_USE_NATIVE_C10D_FUNCTIONAL` that tells `_functional_collective` to use native `c10d_functional` ops. The Python version and the native version will co-exist until we completely switch to the native version after more testing and verification.

NOTE: `DeviceMesh` support for native `c10d_functional` will be added in a subsequent PR.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

Commit:
Allow using native c10d_functional via _functional_collectives

[ghstack-poisoned]

Commit:
Update on "Allow using native c10d_functional via _functional_collectives"




[ghstack-poisoned]

Commit:
Update on "Allow using native c10d_functional via _functional_collectives"




[ghstack-poisoned]

Commit:
Update on "Allow using native c10d_functional via _functional_collectives"




[ghstack-poisoned]

Commit:
Update on "Allow using native c10d_functional via _functional_collectives"




[ghstack-poisoned]

Commit:
Update on "Allow using native c10d_functional via _functional_collectives"




[ghstack-poisoned]

wanchaol commented on torch/distributed/_functional_collectives.py:
```diff
@@ -247,8 +293,23 @@ def all_reduce_coalesced(self: List[torch.Tensor], reduceOp: str, group: RANK_TY
-    tensor_list = torch.ops.c10d_functional.all_reduce_coalesced(self, reduceOp, tag, rankset, group_size)  # type: ignore[attr-defined]
+    if USE_NATIVE_C10D_FUNCTIONAL:
+        assert isinstance(group, str)
+        assert tag == ""
```
why we don't need to call `expand_group_for_native_ops` for allreduce_coalesced?

yifuwang replied:
Good catch. I overlooked this

wanchaol commented on torch/distributed/_functional_collectives.py:
```diff
@@ -379,13 +488,14 @@ def __new__(cls, elem: torch.Tensor):
             device=elem.device, requires_grad=False
         )
         r.elem = elem
+        r.completed = False
```
hmmm I think if the ACT is completed, wouldn't it make more sense to just return the underlying tensor to the user instead of still trying to wrap the data with ACT?

yifuwang replied:
> just return the underlying tensor to the user instead of still trying to wrap the data with ACT

Yeah we should do that. I will update the PR.

The flag still serves some purposes though:
- Verifying the behavior of ACT in unit tests.
- Reduce the number of unneccessary `wait_tensor` calls being added to graph during tracing (when the ACT is being consumed by multiple ops).

wanchaol commented on torch/distributed/_functional_collectives.py:
```diff
@@ -395,12 +505,14 @@ def __tensor_unflatten__(inner_tensors, meta):
     def trigger_wait(self):
-        wait_tensor(self.elem)
-        return self
+        if not self.completed:
```
I feel we should just delete the `trigger_wait` call, I don't actually quite see the value of this API instead of just calling `wait_tensor`?

wanchaol commented on torch/distributed/_functional_collectives.py:
```diff
@@ -507,6 +619,30 @@ def cast_listint(x):
+    elif isinstance(group, str):
+        return group, torch._C._distributed_c10d._resolve_process_group(group).size()
+    else:
+        # TODO(yifu): DeviceMesh supported will be added in a subsequent PR
```
I think device mesh support is actually quite simple to add?
i.e. we can do `device_mesh.get_dim_groups()` and then we will get the process group, then we can handle it the same way as `ProcessGroup`, the complexity here is that inside `get_dim_groups` of DeviceMesh, it still uses the [find_pg_by_tags_and_rank](https://github.com/pytorch/pytorch/blob/main/torch/distributed/_device_mesh.py#L348) to search for a PG in the c10d's global map in Python, wondering if you have ideas how to do this PG resolution in the future?

yifuwang replied:
Currently `_dim_group_infos` tracks `tag` and `rank_set`. I planned to add `group_name` in addition (and eventually remove `tag` and `rank_set`).

wanchaol commented on torch/distributed/_functional_collectives_impl.py:
```diff
@@ -87,6 +90,9 @@ def _wait_reg_dec(ptr, wait_reg):
 
 def _register_tensor_wrapper(tensor) -> None:
+    if USE_NATIVE_C10D_FUNCTIONAL:
+        # Tensor storage -> work mapping is maintained in C++
```
nice!

wanchaol commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1393,6 +1396,7 @@ def _new_process_group_helper(
     _world.pg_map[pg] = (backend, prefix_store)
     _world.pg_names[pg] = group_name
     pg._set_group_name(group_name)
```
can be done in follow up PRs, but I think maybe we can combine `pg._set_group_name` and `register_process_group` together, as I don't quite remember the usage of the former API.

wanchaol commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1451,6 +1455,7 @@ def destroy_process_group(group: Optional[ProcessGroup] = None):
         _world.tags_to_pg.clear()
         _world.pg_coalesce_state.clear()
         _world.pg_default_device.clear()
+        _unregister_all_process_groups()
```
So IIUC, once we have more verification and testing, we can remove the whole python layer global map and universally use C++ global map, and then we can get rid of the double state maintainace across our codebase, is this right?

yifuwang replied:
Yes exactly

wanchaol reviewed (commented):
Nice change! I have some questions inlined and also want to understand more what's the reason we have the `completed` flag inside ACT

Commit:
Update on "Allow using native c10d_functional via _functional_collectives"


This diff introduces an env var `_USE_NATIVE_C10D_FUNCTIONAL` that tells `_functional_collective` to use native `c10d_functional` ops. The Python version and the native version will co-exist until we completely switch to the native version after more testing and verification.

NOTE: `DeviceMesh` support for native `c10d_functional` will be added in a subsequent PR.

[ghstack-poisoned]

Commit:
Update on "Allow using native c10d_functional via _functional_collectives"


This diff introduces an env var `_USE_NATIVE_C10D_FUNCTIONAL` that tells `_functional_collective` to use native `c10d_functional` ops. The Python version and the native version will co-exist until we completely switch to the native version after more testing and verification.

NOTE: `DeviceMesh` support for native `c10d_functional` will be added in a subsequent PR.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

Commit:
Update on "Allow using native c10d_functional via _functional_collectives"


This diff introduces an env var `_USE_NATIVE_C10D_FUNCTIONAL` that tells `_functional_collective` to use native `c10d_functional` ops. The Python version and the native version will co-exist until we completely switch to the native version after more testing and verification.

NOTE: `DeviceMesh` support for native `c10d_functional` will be added in a subsequent PR.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

Commit:
Update on "Allow using native c10d_functional via _functional_collectives"


This diff introduces an env var `_USE_NATIVE_C10D_FUNCTIONAL` that tells `_functional_collective` to use native `c10d_functional` ops. The Python version and the native version will co-exist until we completely switch to the native version after more testing and verification.

NOTE: `DeviceMesh` support for native `c10d_functional` will be added in a subsequent PR.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

Commit:
Update on "Allow using native c10d_functional via _functional_collectives"


This diff introduces an env var `_USE_NATIVE_C10D_FUNCTIONAL` that tells `_functional_collective` to use native `c10d_functional` ops. The Python version and the native version will co-exist until we completely switch to the native version after more testing and verification.

NOTE: `DeviceMesh` support for native `c10d_functional` will be added in a subsequent PR.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

LucasLLC reviewed (approved):
lgtm

Commit:
Update on "Allow using native c10d_functional via _functional_collectives"


This diff introduces an env var `_USE_NATIVE_C10D_FUNCTIONAL` that tells `_functional_collective` to use native `c10d_functional` ops. The Python version and the native version will co-exist until we completely switch to the native version after more testing and verification.

NOTE: `DeviceMesh` support for native `c10d_functional` will be added in a subsequent PR.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

wconstab reviewed (changes requested):
looks good, just a couple questions
- what is unregister_process_group for? maybe it should be a separate PR but i would be satisfied with even an explanation or some docs

- it looks like you're planning to do an actual rollout of a feature, which makes me less against adding a new ENV.  We don't want to add envs/modes.. But for a rollout this is ok.  You can follow https://github.com/pytorch/pytorch/pull/118451 to set up a JK knob for your env and control the internal rollout.  We should set a plan for how long it takes and ensuring the old path is cleaned up.

Commit:
Update on "Allow using native c10d_functional via _functional_collectives"


This diff introduces an env var `_USE_NATIVE_C10D_FUNCTIONAL` that tells `_functional_collective` to use native `c10d_functional` ops. The Python version and the native version will co-exist until we completely switch to the native version after more testing and verification.

NOTE: `DeviceMesh` support for native `c10d_functional` will be added in a subsequent PR.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

yifuwang commented:
> * what is unregister_process_group for? maybe it should be a separate PR but i would be satisfied with even an explanation or some docs

@wconstab I added some comments on the pybind.

> it looks like you're planning to do an actual rollout of a feature, which makes me less against adding a new ENV. We don't want to add envs/modes.. But for a rollout this is ok. You can follow https://github.com/pytorch/pytorch/pull/118451 to set up a JK knob for your env and control the internal rollout. We should set a plan for how long it takes and ensuring the old path is cleaned up.

Yeah this env var isn't meant for users and will be removed once the migration completes. A jk knob for rollout sounds good and I will do that. Right now we are not there yet. This env var helps me perform manual validations, and it allows me to run unit tests that use functional collectives with both the existing impl and the new impl in CI, while I close the functionality gap.

fegin commented:
@yifuwang This is nice. If I would like to test it with DDP compilation, I would also need to include https://github.com/pytorch/pytorch/pull/118498, is that correct?

yifuwang commented:
> If I would like to test it with DDP compilation, I would also need to include https://github.com/pytorch/pytorch/pull/118498, is that correct?

Hey @fegin, I think this PR alone should be enough. The PR you mentioned is more on the optimization side of things.

wconstab reviewed (approved):
lgtm, suggest also updating PR desc so its in the merge commit, with some summary of 'add new "native registry" that is for c++/py, plan to deprecate py one' or similar

wanchaol commented on torch/distributed/_functional_collectives_impl.py:
```diff
@@ -92,6 +95,9 @@ def _wait_reg_dec(ptr, wait_reg):
 def _register_tensor_wrapper(tensor) -> None:
+    if USE_NATIVE_C10D_FUNCTIONAL:
+        # Tensor storage -> work mapping is maintained in C++
+        return
```
Question: how we are maintaining the ACT wrapper subclass instance mapping to the work? 

iiuc the current behavior we have:
```
act_tensor = funcol.allreduce(..) # the returned act_tensor.elem map to work, along with `act_tensor` itself
new_act = act_tensor.view(..) # the returned new_act also tracks map to work
```
Our new behavior is that we don't track the tensor wrapper subclass mapping to the work, and we only track the underlying `act.elem` mapping to the work?

yifuwang replied:
> Our new behavior is that we don't track the tensor wrapper subclass mapping to the work, and we only track the underlying act.elem mapping to the work?

Hey @wanchaol, in the new impl, we use `tensor.storage().getWeakStorageImpl()` to identify the associated work. The scenario you describe should work. This is a very useful point. I will add a test case for this scenario!

wanchaol reviewed (approved):
lgtm! one question about our new behavior of tensor/work mapping

yifuwang commented:
@pytorchbot merge
